presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java

```diff
@@ -30,70 +39,21 @@
 
     private GroupingOperationFunction() {}
 
-    /**
-     * The grouping function is used in conjunction with GROUPING SETS, ROLLUP and CUBE to
-     * indicate which columns are present in that grouping.
-     *
-     * <p>The grouping function must be invoked with arguments that exactly match the columns
-     * referenced in the corresponding GROUPING SET, ROLLUP or CUBE clause at the associated
-     * query level. Those column arguments are not evaluated and instead the function is
-     * re-written with the arguments below.
-     *
-     * <p>To compute the resulting bit set for a particular row, bits are assigned to the
-     * argument columns with the rightmost column being the most significant bit. For a
-     * given grouping, a bit is set to 0 if the corresponding column is included in the
-     * grouping and 1 otherwise. For an example, see the SQL documentation for the
-     * function.
-     *
-     * @param groupId An ordinal indicating which grouping is currently being processed.
-     *        Each grouping is assigned a unique monotonically increasing integer.
-     * @param columns The column arguments with which the function was
-     *        invoked converted to ordinals with respect to the base table column
-     *        ordering.
-     * @param groupingSetDescriptors A collection of ordinal lists where the index of
-     *        the list is the groupId and the list itself contains the ordinals of the
-     *        columns present in the grouping. For example: [[0, 2], [2], [0, 1, 2]]
-     *        means the the 0th list contains the set of columns that are present in
-     *        the 0th grouping.
-     * @return A bit set converted to decimal indicating which columns are present in
-     *         the grouping. If a column is NOT present in the grouping its corresponding
-     *         bit is set to 1 and to 0 if the column is present in the grouping.
-     */
     @ScalarFunction(value = INTEGER_GROUPING, deterministic = false)
     @SqlType(StandardTypes.INTEGER)
     public static long integerGrouping(
-            @SqlType(BIGINT) long groupId,
-            @SqlType("ListLiteral") List<Integer> columns,
-            @SqlType("ListLiteral") List<List<Integer>> groupingSetDescriptors)
+            @SqlType("array(integer)") Block array,
+            @SqlType(BIGINT) long groupId)
     {
-        return calculateGrouping(groupId, columns, groupingSetDescriptors);
+        return longElementAt(IntegerType.INTEGER, array, groupId);
```

Comment:
The reason I did it this way is that otherwise the `element_at` function gets pushed down when `grouping` is in a predicate. For example the following query fails:
```
SELECT a, b, sum(c), grouping(a, b)
FROM (VALUES ('h', 'j', 11), ('k', 'l', 7) ) AS t (a, b, c)
GROUP BY GROUPING SETS ( (a), (b))
HAVING grouping(a, b) > 1 
```
I had a look around in `PredicatePushDown` and couldn't find an elegant solution. I can add a total hack to `DeterminismEvaluator` to mark `element_at` instances containing a `grouping` symbol as non-deterministic but I feel like the current solution is probably better. I'm open to suggestions for  a better solution.

Reply:
Be sure to add such a query to tests (unless it's already there)

> I can add a total hack to DeterminismEvaluator to mark element_at

Having deterministic function with `@ScalarFunction(deterministic = false)` is also a hack...  I'd like to understand why wrong push down takes place (but i can't dig into this right now)

Reply:
> Be sure to add such a query to tests (unless it's already there)

Already a test.

> Having deterministic function with @ScalarFunction(deterministic = false) is also a hack

I think in the context of Presto a non deterministic function is a function that should not be pushed down (in addition to the more straightforward meaning of deterministic). https://github.com/prestodb/presto/pull/5465#r70119024. I could be wrong though but that is my understanding.

>  I'd like to understand why wrong push down takes place

Let me try to explain what happens. As the `PredicatePushDown` visitor traverses the plan, it appends whatever predicate is present in a `FilterNode` and passes it down via the context. However, in `PredicatePushDown#visitAggregation()` all expressions that are not deterministic get removed. Determinism is evaluated in the `DeterminismEvaluator` class which only overrides the traversal method for `FunctionCall` and does a simple name comparison. That is where I originally hard coded the grouping function names (now realizing that the `@ScalarFunction(deterministic = false)` is made obsolete by these lines) so they get treated as not deterministic and get pruned.

However, with `element_at`, the function is marked as being deterministic and is further pushed down. Once the function reaches the `PredicatePushDown#visitGroupId` all symbols are extracted to ensure that no groupId symbols get pushed below the `GroupId` node.

A possible solution would be to edit `DeterminismEvaluator` to traverse the plan and extract any groupid symbols from child nodes and see if any of those symbols are present in the function. If that is the case then mark the function as being non-deterministic.

Reply:
I think I have problem with understanding what "deterministic" means. For example `substr(x, 1)` is deterministic and `random()`, `random(n)` are not.  For example, `SortExpressionExtractor` uses the notion of determinism to determine whether an `expression < symbol` is suitable for binary search over symbol space. From that point of view, `grouping(..)` is as deterministic as `substr(..)`, isn't it?

Reply:
To be 100% explicit I believe deterministic means 1) always returns the same output given the same input and 2) can be pushed down. `substr()` fulfills both of these categories. `random()` does not fulfill 1) and looking at `DeterminismEvaluator` does not fulfill 2) although I am unclear as to why that's the case.

> From that point of view, grouping(..) is as deterministic as substr(..), isn't it?

So `grouping` fulfills 1) but does not fulfill 2). I agree with you that having a "dual" meaning for deterministic is confusing (if that is indeed the case). So maybe we need to introduce a new concept of a 'not-pushable' function? However that would only apply to `grouping` so maybe such a concept shouldn't be introduced until we have more examples? A stop gap solution in the meantime could be what I suggest above. I am of course always open to other ideas.

Reply:
As far as I can tell, the issue is that PredicatePushDown has an artificial limitation on predicates not being able to depend directly on the groupid column. This was a protection we added because it could never happen in practice, but there's no reason this shouldn't be allowed.

I'm reworking this PR to relax that constraint and not need those artificial functions.